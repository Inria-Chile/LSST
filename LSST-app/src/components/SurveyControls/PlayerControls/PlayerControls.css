
.player-controls {
    width: 45%;
}

.PlaybackControls {
    text-align: center;
}