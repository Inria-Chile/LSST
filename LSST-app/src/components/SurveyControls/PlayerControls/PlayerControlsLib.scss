
$icon-size: 20px;

$green: #2d3d4e;
$white: #aab1b9;
$gray: #111;
$progress-bkg: #090;
$elapsed-bkg: #fff;
// $color-hover: saturate(lighten($green, 10%), 20%);
// $color-active: desaturate(darken($green, 20%), 20%);
$color-hover: desaturate(darken($green, 20%), 10%);
$color-active: fade-out($color-hover, 0.5);


//
// Buttons
//

.PlayButton,
.PauseButton,
.PrevButton,
.NextButton,
.SoundOnButton,
.SoundOffButton {
  margin-right: 10px;
  padding: 8px;
  background: $white;
  border: 2px solid $green;
  border-radius: 100%;
  outline: none;
  line-height: 0.5;

  &[disabled] {
    opacity: 0.5;
  }

  &.isEnabled {
    cursor: pointer;

    &:hover {
      border-color: $color-hover;

      .Icon-shape {
        fill: $color-hover;
      }
    }

    &:active {
      border-color: $color-active;

      .Icon-shape {
        fill: $color-active;
      }
    }
  }
}

.Icon {
  width: $icon-size;
  height: $icon-size;
}

.Icon-shape {
  fill: $green;
}


//
// Sliders
//

.ProgressBar,
.VolumeSlider {
  position: relative;
  background: #aab1b9;
  border-radius: 4px;

  &-elapsed,
  &-value {
    background: $progress-bkg;
    border-radius: 4px;
  }

  &-seek {
    position: absolute;
    cursor: pointer;
  }

  &-handle {
    position: absolute;
    width: 26px;
    height: 26px;
    background: $elapsed-bkg;
    border-radius: 100%;
    transform: scale(0);
    transition: transform 0.2s;
  }

  &.isSeekable,
  &.isEnabled {
      .ProgressBar-handle,
      .VolumeSlider-handle {
        transform: scale(1);
        background-color: #aab1b9;
        border: 5px solid white;        
      }

      .ProgressBar-intent,
      .VolumeSlider-intent {
        position: absolute;
        background: rgba(0,0,0,0.05);
      }

    .ProgressBar-elapsed,
    .VolumeSlider-value {
      background: $elapsed-bkg;
    }
  }
}

.ProgressBar {
  height: 7px;
  margin: 10px 0;

  &-elapsed {
    height: 150%;
    transform: translateY(-15%);
  }

  &-seek {
    top: -6px;
    right: 0;
    bottom: -6px;
    left: 0;
  }

  &-handle {
    top: 0;
    margin: -8px 0 0 -12px;
  }

  &.isSeekable {
    &:hover {
      .ProgressBar-handle {
        transform: scale(1);
      }

      .ProgressBar-intent {
        top: 0;
        left: 0;
        height: 100%;
      }
    }
  }

  &.isLoading {
    opacity: 0.5;
    background: #ccc;
    animation: barLoading 2s infinite;
  }
}

.VolumeSlider {
  width: 8px;
  height: 100px;
  // margin: 10px 0;

  &-value {
    position: absolute;
    bottom: 0;
    width: 100%;
  }

  &-seek {
    left: -6px;
    top: 0;
    right: -6px;
    bottom: 0;
  }

  &-handle {
    left: 0;
    margin: 0 0 -8px -4px;
  }

  &.isEnabled {
    &:hover {
      .VolumeSlider-handle {
        transform: scale(1);
      }

      .VolumeSlider-intent {
        bottom: 0;
        left: 0;
        width: 100%;
      }
    }
  }
}