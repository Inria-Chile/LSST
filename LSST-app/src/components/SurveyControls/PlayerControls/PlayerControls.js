import React, { Component } from 'react';
// import ReactDOM from 'react-dom';
// import './SurveyControls.css';
import './PlayerControls.css';
import './PlayerControlsLib.css';
import { PlaybackControls, ProgressBar } from 'react-player-controls'

class PlayerControls extends Component {


    constructor() {
        super();
        this.playbackTimerId = null;
        this.state = {
            totalTime: 190,
            currentTime: Infinity,
            isSeekable: true,
            isPlayable: true,
            hasPrevious: true,
            hasNext: true,
        };
    }

    playbackChange = (isPlaying) => {
        // console.log("playbackChange");
        let currentTime = this.state.currentTime;
        if(isPlaying){
            if(this.state.currentTime === this.props.endDate || this.state.currentTime === Infinity)
                currentTime = 0;
            // this.playbackTimerId = setInterval(this.animate, 300);
        }else
            this.stopAnimating();
        this.setState({ ...this.state, isPlaying, currentTime: currentTime })
    }

    showPrevious = () => {
        // console.log('showPrevious', this.props.startDate, this.props.endDate, Math.min(this.props.endDate, this.state.currentTime));
        this.setState({currentTime: 0});
        this.props.setDisplayedDateLimits(new Date(this.props.startDate), new Date(this.props.startDate));
    }

    showNext = () => {
        // console.log('showNext', this.props.startDate, this.props.endDate, Math.min(this.props.endDate, this.state.currentTime));
        this.setState({currentTime: this.props.endDate});
        this.props.setDisplayedDateLimits(new Date(this.props.startDate), new Date(this.props.endDate));
    }

    seekEnd = (time) => {
        this.animate();
    }

    getTimeInterval = () => {
        return this.props?(this.props.endDate - this.props.startDate)/100 : 1000;
    }

    animate = () => {
        if(this.props.startDate === null || this.props.endDate === null || this.state.currentTime > this.props.endDate){
            this.setState({currentTime: this.props.endDate});
            this.stopAnimating();
            return;            
        }
        // console.log('animating', this.props.startDate, this.props.endDate, Math.min(this.props.endDate, this.state.currentTime + this.getTimeInterval()));
        this.setState({currentTime: this.state.currentTime + this.getTimeInterval()});
        this.props.setDisplayedDateLimits(new Date(this.props.startDate), new Date(Math.min(this.props.endDate, this.props.startDate + this.state.currentTime + this.getTimeInterval())));
    }

    stopAnimating = () => {
        clearInterval(this.playbackTimerId);
        this.playbackTimerId = null;
        this.setState({isPlaying: false});
    }
    
    componentDidUpdate(prevProps, prevState){
        if((this.state.isPlaying && !prevState) || (this.state.isPlaying && !prevState.isPlaying))
            this.playbackTimerId = setInterval(this.animate, 300);
        if(prevProps.endDate !== this.props.endDate)
            this.setState({currentTime: this.props.endDate});
    }

    render() {
        return (
            <div className="player-controls">
                <PlaybackControls
                    isPlayable={this.state.isPlayable}
                    isPlaying={this.state.isPlaying}
                    showPrevious={this.state.showPrevious}
                    hasPrevious={this.state.hasPrevious}
                    showNext={this.state.showNext}
                    hasNext={this.state.hasNext}
                    onPlaybackChange={isPlaying => this.playbackChange(isPlaying)}
                    onPrevious={this.showPrevious}
                    onNext={this.showNext}
                />

                <ProgressBar
                    totalTime={this.props.endDate - this.props.startDate}
                    currentTime={this.state.currentTime}
                    isSeekable={this.state.isSeekable}
                    onSeek={time => this.setState(() => ({ currentTime: time }))}
                    onSeekStart={time => this.setState(() => ({ lastSeekStart: time }))}
                    onSeekEnd={this.seekEnd}
                    onIntent={time => this.setState(() => ({ lastIntent: time }))}
                />
            </div>
        );
    }
}

export default PlayerControls;