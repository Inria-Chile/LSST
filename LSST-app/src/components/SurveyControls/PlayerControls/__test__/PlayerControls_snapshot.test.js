import React from 'react';
import PlayerControls from '../PlayerControls';
import renderer from 'react-test-renderer';
import shallowRenderer from 'react-test-renderer/shallow';
import moment from 'moment';


/*---- Snapshot testing ----*/
describe('PlayerControls_Test',function(){
    
  
  it('renders correctly with shallowRender',()=> {
    let setDisplayedDateLimits = jest.fn();
    const renderer = new shallowRenderer();
    const tree = renderer.render(
        <PlayerControls startDate={moment(moment("1994-01-01"))} 
        endDate={moment("1994-01-01")} 
        setDisplayedDateLimits={setDisplayedDateLimits}/>);
      expect(tree).toMatchSnapshot();
  });


})
