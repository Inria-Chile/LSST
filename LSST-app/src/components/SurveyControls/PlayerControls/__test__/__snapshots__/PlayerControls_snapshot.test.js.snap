// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerControls_Test renders correctly with shallowRender 1`] = `
<div
  className="player-controls"
>
  <PlaybackControls
    childClasses={Object {}}
    childrenStyles={Object {}}
    className="PlaybackControls"
    extraClasses=""
    hasNext={true}
    hasPrevious={true}
    isPlayable={true}
    isPlaying={false}
    onNext={[Function]}
    onPlaybackChange={[Function]}
    onPrevious={[Function]}
    showNext={true}
    showPrevious={true}
    style={Object {}}
  />
  <ProgressBar
    childClasses={Object {}}
    childrenStyles={Object {}}
    className="ProgressBar"
    currentTime={Infinity}
    extraClasses=""
    isSeekable={true}
    onIntent={[Function]}
    onSeek={[Function]}
    onSeekEnd={[Function]}
    onSeekStart={[Function]}
    style={Object {}}
    totalTime={0}
  />
</div>
`;
